somewhat surprising match default binding interaction

It took me a while to figure out the compilation error in this example ([try it yourself](https://play.rust-lang.org/?gist=4b9f2f8b9dd7f1bfb4219783ff5ad0c3&version=nightly)):

```rust
#![feature(match_default_bindings)]

fn surprise(x: i32) { }

fn main() {
  let x = &(1, &2);
  let (_, &b) = x;
  surprise(b);
}
```

Here you get:

```
error[E0308]: mismatched types
 --> src/main.rs:9:12
  |
9 |   surprise(b);
  |            ^
  |            |
  |            expected i32, found &{integer}
  |            help: consider dereferencing the borrow: `*b`
  |
  = note: expected type `i32`
             found type `&{integer}`
```

Naturally I tried changing the pattern `&&b`. That gives you:

```
error[E0308]: mismatched types
 --> src/main.rs:7:12
  |
7 |   let (_, &&b) = x;
  |            ^^ expected integral variable, found reference
  |
  = note: expected type `{integer}`
             found type `&_`
  = help: did you mean `b: &{integer}`?
```

What the heck? 

What's going on here is that the filter is giving us an `&(i32, &i32)`. When we skip the first `&`, we get into "ref by default" mode, but when we explicitly acknowledge the second one, we do *not* get back into "by value" mode. That's kind of annoying.

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -151,6 +151,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     err.emit();
                 }
             }
+        } else if let PatKind::Ref(..) = pat.node {
+            // When you encounter a `&pat` pattern, reset to "by
+            // value". This is so that `x` and `y` here are by value,
+            // as they appear to be:
+            //
+            // ```
+            // match &(&22, &44) {
+            //   (&x, &y) => ...
+            // }
+            // ```
+            //
+            // cc #46688
+            def_bm = ty::BindByValue(hir::MutImmutable);
         }
 
         // Lose mutability now that we know binding mode and discriminant type.

--- a/src/test/run-pass/rfc-2005-default-binding-mode/reset-mode.rs
+++ b/src/test/run-pass/rfc-2005-default-binding-mode/reset-mode.rs
@@ -1,0 +1,25 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(match_default_bindings)]
+
+// Test that we "reset" the mode as we pass through a `&` pattern.
+//
+// cc #46688
+
+fn surprise(x: i32) {
+    assert_eq!(x, 2);
+}
+
+fn main() {
+    let x = &(1, &2);
+    let (_, &b) = x;
+    surprise(b);
+}

